File: include/extras-ktexteditor/libtexthintinterface.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");
const std = @import("std");

/// https://api.kde.org/ktexteditor-texthintprovider.html
pub const ktexteditor__texthintprovider = struct {
    /// New constructs a new KTextEditor::TextHintProvider object.
    ///
    ///
    pub fn New() QtC.KTextEditor__TextHintProvider {
        return qtc.KTextEditor__TextHintProvider_new();
    }

    /// [Qt documentation](https://api.kde.org/ktexteditor-texthintprovider.html#textHint)
    ///
    /// ``` self: QtC.KTextEditor__TextHintProvider, view: QtC.KTextEditor__View, position: QtC.KTextEditor__Cursor, allocator: std.mem.Allocator ```
    pub fn TextHint(self: ?*anyopaque, view: ?*anyopaque, position: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.KTextEditor__TextHintProvider_TextHint(@ptrCast(self), @ptrCast(view), @ptrCast(position));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("ktexteditor::texthintprovider.TextHint: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// [Qt documentation](https://api.kde.org/ktexteditor-texthintprovider.html#textHint)
    ///
    /// Allows for overriding the related default method
    ///
    /// ``` self: QtC.KTextEditor__TextHintProvider, callback: *const fn (self: QtC.KTextEditor__TextHintProvider, view: QtC.KTextEditor__View, position: QtC.KTextEditor__Cursor) callconv(.c) [*:0]const u8 ```
    pub fn OnTextHint(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque, ?*anyopaque) callconv(.c) [*:0]const u8) void {
        qtc.KTextEditor__TextHintProvider_OnTextHint(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// [Qt documentation](https://api.kde.org/ktexteditor-texthintprovider.html#textHint)
    ///
    /// Base class method implementation
    ///
    /// ``` self: QtC.KTextEditor__TextHintProvider, view: QtC.KTextEditor__View, position: QtC.KTextEditor__Cursor, allocator: std.mem.Allocator ```
    pub fn QBaseTextHint(self: ?*anyopaque, view: ?*anyopaque, position: ?*anyopaque, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.KTextEditor__TextHintProvider_QBaseTextHint(@ptrCast(self), @ptrCast(view), @ptrCast(position));
        defer qtc.libqt_string_free(&_str);
        const _ret = allocator.alloc(u8, _str.len) catch @panic("ktexteditor::texthintprovider.TextHint: Memory allocation failed");
        @memcpy(_ret, _str.data[0.._str.len]);
        return _ret;
    }

    /// Delete this object from C++ memory.
    ///
    /// ``` self: QtC.KTextEditor__TextHintProvider ```
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.KTextEditor__TextHintProvider_Delete(@ptrCast(self));
    }
};

File: src_original/extras-ktexteditor/libtexthintinterface.cpp
#include <KTextEditor/Cursor>
#define WORKAROUND_INNER_CLASS_DEFINITION_KTextEditor__TextHintProvider
#define WORKAROUND_INNER_CLASS_DEFINITION_KTextEditor__View
#include <QString>
#include <QByteArray>
#include <cstring>
#include <texthintinterface.h>
#include "libtexthintinterface.h"
#include "libtexthintinterface.hxx"

KTextEditor__TextHintProvider* KTextEditor__TextHintProvider_new() {
    return new VirtualKTextEditorTextHintProvider();
}

libqt_string KTextEditor__TextHintProvider_TextHint(KTextEditor__TextHintProvider* self, KTextEditor__View* view, const KTextEditor__Cursor* position) {
    auto* vktexteditor__texthintprovider = dynamic_cast<VirtualKTextEditorTextHintProvider*>(self);
    if (vktexteditor__texthintprovider && vktexteditor__texthintprovider->isVirtualKTextEditorTextHintProvider) {
        QString _ret = vktexteditor__texthintprovider->textHint(view, *position);
        // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
        QByteArray _b = _ret.toUtf8();
        libqt_string _str;
        _str.len = _b.length();
        _str.data = static_cast<const char*>(malloc(_str.len + 1));
        memcpy((void*)_str.data, _b.data(), _str.len);
        ((char*)_str.data)[_str.len] = '\0';
        return _str;
    } else {
        QString _ret = ((VirtualKTextEditorTextHintProvider*)self)->textHint(view, *position);
        // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
        QByteArray _b = _ret.toUtf8();
        libqt_string _str;
        _str.len = _b.length();
        _str.data = static_cast<const char*>(malloc(_str.len + 1));
        memcpy((void*)_str.data, _b.data(), _str.len);
        ((char*)_str.data)[_str.len] = '\0';
        return _str;
    }
}

// Base class handler implementation
libqt_string KTextEditor__TextHintProvider_QBaseTextHint(KTextEditor__TextHintProvider* self, KTextEditor__View* view, const KTextEditor__Cursor* position) {
    auto* vktexteditortexthintprovider = dynamic_cast<VirtualKTextEditorTextHintProvider*>(self);
    if (vktexteditortexthintprovider && vktexteditortexthintprovider->isVirtualKTextEditorTextHintProvider) {
        vktexteditortexthintprovider->setKTextEditor__TextHintProvider_TextHint_IsBase(true);
        QString _ret = vktexteditortexthintprovider->textHint(view, *position);
        // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
        QByteArray _b = _ret.toUtf8();
        libqt_string _str;
        _str.len = _b.length();
        _str.data = static_cast<const char*>(malloc(_str.len + 1));
        memcpy((void*)_str.data, _b.data(), _str.len);
        ((char*)_str.data)[_str.len] = '\0';
        return _str;
    } else {
        QString _ret = ((VirtualKTextEditorTextHintProvider*)self)->textHint(view, *position);
        // Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
        QByteArray _b = _ret.toUtf8();
        libqt_string _str;
        _str.len = _b.length();
        _str.data = static_cast<const char*>(malloc(_str.len + 1));
        memcpy((void*)_str.data, _b.data(), _str.len);
        ((char*)_str.data)[_str.len] = '\0';
        return _str;
    }
}

// Auxiliary method to allow providing re-implementation
void KTextEditor__TextHintProvider_OnTextHint(KTextEditor__TextHintProvider* self, intptr_t slot) {
    auto* vktexteditortexthintprovider = dynamic_cast<VirtualKTextEditorTextHintProvider*>(self);
    if (vktexteditortexthintprovider && vktexteditortexthintprovider->isVirtualKTextEditorTextHintProvider) {
        vktexteditortexthintprovider->setKTextEditor__TextHintProvider_TextHint_Callback(reinterpret_cast<VirtualKTextEditorTextHintProvider::KTextEditor__TextHintProvider_TextHint_Callback>(slot));
    }
}

void KTextEditor__TextHintProvider_Delete(KTextEditor__TextHintProvider* self) {
    delete self;
}
